### ISSUE TITLE
Out of memory unpacking Requests

### ISSUE THREAD

**Issue by battis** [instructure/canvas-data-cli#8]
I'm running this on my laptop (16GB RAM), and keep being told that, when I unpack Requests, that Javascript is running out of memory. I note that my requests.txt file went from 825K to ~45MB as a result of this unpacking, so I'm not _sure_ that it didn't get fully unpacked. But…

```
$ canvasDataCLI unpack -c config.js -f requests
outputting requests to /Users/sethbattis/canvas-data-cli/unpackedFiles/requests.txt

<--- Last few GCs --->

  440187 ms: Mark-sweep 1390.3 (1434.8) -> 1390.2 (1434.8) MB, 5311.6 / 0 ms [allocation failure] [GC in old space requested].
  444771 ms: Mark-sweep 1390.2 (1434.8) -> 1390.2 (1434.8) MB, 4582.4 / 0 ms (+ 1.2 ms in 1 steps since start of marking, biggest step 1.2 ms) [allocation failure] [GC in old space requested].
  449126 ms: Mark-sweep 1390.2 (1434.8) -> 1390.2 (1434.8) MB, 4354.9 / 0 ms [last resort gc].
  453310 ms: Mark-sweep 1390.2 (1434.8) -> 1390.2 (1434.8) MB, 4184.3 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x14d6bb8c9e31 <JS Object>
    2: onwrite(aka onwrite) [_stream_writable.js:~329] [pc=0x1aab841dec64] (this=0x14d6bb804189 <undefined>,stream=0x33b49e8f05b9 <a WriteStream with map 0x3949478aa079>,er=0x14d6bb804189 <undefined>)
    3: /* anonymous */(aka /* anonymous */) [_stream_writable.js:~88] [pc=0x1aab84113e97] (this=0x14d6bb804189 <undefined>,er=0x14d6bb804189 <undefined>)
    4: arguments adaptor frame: 0->1
    5...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [/usr/local/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 5: v8::internal::Factory::NewFixedArray(int, v8::internal::PretenureFlag) [/usr/local/bin/node]
 6: v8::internal::TypeFeedbackVector::New(v8::internal::Isolate*, v8::internal::Handle<v8::internal::TypeFeedbackMetadata>) [/usr/local/bin/node]
 7: v8::internal::CompilationInfo::EnsureFeedbackVector() [/usr/local/bin/node]
 8: v8::internal::FullCodeGenerator::MakeCode(v8::internal::CompilationInfo*) [/usr/local/bin/node]
 9: v8::internal::GenerateBaselineCode(v8::internal::CompilationInfo*) [/usr/local/bin/node]
10: v8::internal::CompileBaselineCode(v8::internal::CompilationInfo*) [/usr/local/bin/node]
11: v8::internal::GetUnoptimizedCodeCommon(v8::internal::CompilationInfo*) [/usr/local/bin/node]
12: v8::internal::Compiler::GetLazyCode(v8::internal::Handle<v8::internal::JSFunction>) [/usr/local/bin/node]
13: v8::internal::Runtime_CompileLazy(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
14: 0x1aab83d0961b
15: 0x1aab83d31ed9
Abort trap: 6
```

**Comment by battis** [comment]
Hokay… it doesn't seem, after a casual perusal of the documentation, like there's any particularly good way of configuring node's memory usage as part of the make process of a CLI app. Or when invoking the app itself, after it has been compiled. I could well be wrong, but that's my impression.

After digging around for a while, struggling to figure out how to trick node into using more memory, I discovered [this delightful tidbit](http://sambal.org/2014/02/passing-options-node-shebang-line/) that suggests a way to pass parameters to node within a node executable file. It's really pretty slick.

Having installed canvas-data-cli per the readme, I went into executable file (findable using `which canvasDataCli` -- mine is in `/usr/local/bin/canvasDataCli` which is really a symlink to another file buried in deep in the bowels of `/usr/local/bin/`, but whatevs…).

I edited `/usr/local/bin/canvasDataCli` it to read as follows:

```bash
#!/bin/sh 
":" //# comment; exec /usr/bin/env node --max_old_space_size=8192 "$0" "$@"
var cli = require('../').cli
var argv = cli.cli.parse(process.argv.slice(2))
cli.run(argv)
```

This is the trick suggested in the link above, which casts this not as a node executable (as the original shebang did), but as a bash executable, and then allows me to call node with whichever parameters I want -- `max_old_space_size` being _very_ high on my list. I allocated 8GB of memory to the process in this case (8192MB).

> 8GB may not be enough: I just watched this sucker run for about an hour, and it threw in the towel while finishing processing my two largest data files. I'm about to try it again with even more memory allocated, but it does seem like this basically addresses the _first_ problem that I had.
> 
> I've pretty much thrown in the towel on understanding modern memory management -- I watched the virtual memory on this process get to about 30GB, and the real memory hover around 8GB for most of that run.

Nota bene that this hack should work on *nix systems, include macOS where I'm running. There isn't a shebang-equivalent on Windows, but I imagine there is some similar hack there as well.

**Comment by battis** [comment]
One supposes that another, more pragmatic approach would be to gunzip all of the files in `…/dataFiles/requests/`, concatenate them all together (`cat * > requests.txt`) and slap the column headers from schema.json on the top and call it a day.

But I do so want the flippin' script to _work_.

**Comment by battis** [comment]
Right then. Turns out that upgrade node addresses some of the cruft I was seeing, even after expanding the memory allocation to 8GB. In the end, I was running the homebrew install of node 7.2.1 with `--max_old_space_size=10240` (10GB) and, after about an hour and a half… it worked! It was super-convenient to be able to see my memory pressure redlined from across the room to know that everything was still running.

So… yeah. There has _got_ to be a better way.

**Comment by jasonheffner** [comment]
Thanks for the info on where to adjust for the pass-through variables to node. I was trying to send --max-old-space-size=8192 and this helps. I didn't have the chance to figure out where to add this in. I had to break up our unpack into several jobs for all the files since they were hitting the memory limit when done in one command.

As far as requests though I don't extract those since they are far too large. The code was updated sometime last year to process the files for errors which I believe where all the extra overhead was added. It's used to be coded to simply extract, concatenate, and slap the column headers on as you described so well. It's well worth the effort though on my end as my postrges imports are flawless now. 

Request data, at least for our current implementation, I feed the gzip files directly into Splunk which extracts and imports the request data, which is up in the billions of records now. I find it so much easier to query/analyze this data through splunk than tring to write queries in postgres directly.

**Comment by battis** [comment]
I wonder how much of this memory consumption is driven simply by the size of the data files? I'm noticing that, of course, it takes longest to process the largest files. If the Canvas Data API were to deliver requests in smaller chunks, would this be a more manageable task to unpack that data?

**Comment by addisonj** [comment]
Glad to see some progress on at least getting it to work... but yeah, you are totally correct, this really shouldn't be this memory expensive. 

This is being done internally with a node.js streams, which should minimize the amount of memory needed, however, that clearly doesn't seem to be happening quite as expected. It might have to do with large chunk size in zlib, but that still looks like its trying to store most of the file...

I will do a bit of exploration of this and see if I can get a more concrete answer and fix

**Comment by addisonj** [comment]
@battis I think I *might* have a fix here https://github.com/instructure/canvas-data-cli/tree/lazy_stream. I *think* the problem may have been that streams started to buffer data rather that waiting for the previous file to completely process. If you can try with this branch I would love to see if it helps alleviating things:

To try it:
`git clone https://github.com/instructure/canvas-data-cli.git && git checkout lazy_stream && npm install -g .`

**Comment by battis** [comment]
Actually it was more:

`git clone https://github.com/instructure/canvas-data-cli.git && cd canvas-data-cli && git checkout lazy_stream && make installLocal`

…but either way, the change was _astonishing_. It took 6 minutes to unpack requests.txt and never went of 3GB of virtual memory. My fan didn't even turn on.

Seems shiny to me!

**Comment by addisonj** [comment]
Okay, this is now released as `0.5.4`. Closing this, thanks for the help and patience!
